numpy/lib/_npyio_impl.py

```diff
@@ -1559,84 +1559,101 @@ def first_write(self, v):
+            elif isinstance(fstring_fmt, (list, tuple)):
+                raise ValueError(
+                    "Multi-column fstring_fmt is not supported. "
+                    "Provide a single string for fstring_fmt."
+                )
             else:
-                format = fmt
+                raise ValueError("fstring_fmt must be a string.")
+            format_row = delimiter.join(fstring_fmt)
         else:
-            raise ValueError('invalid fmt: %r' % (fmt,))
+            if isinstance(fmt, (list, tuple)):
+                if len(fmt) != ncol:
+                    raise ValueError(f"fmt has wrong shape: {fmt}")
+                format = delimiter.join(fmt)
+            elif isinstance(fmt, str):
+                n_fmt_chars = fmt.count('%')
+                if n_fmt_chars == 1:
+                    fmt = [fmt] * ncol
+                    format = delimiter.join(fmt)
+                elif iscomplex_X and n_fmt_chars != (2 * ncol):
+                    raise ValueError(f"fmt has wrong number of %% formats: {fmt}")
+                elif not iscomplex_X and n_fmt_chars != ncol:
+                    raise ValueError(f"fmt has wrong number of %% formats: {fmt}")
+                else:
+                    format = fmt
+            else:
+                raise ValueError(f"Invalid fmt: {fmt!r}")
 
-        if len(header) > 0:
-            header = header.replace('\n', '\n' + comments)
+        # Write header
+        if header:
+            header = header.replace("\n", "\n" + comments)
             fh.write(comments + header + newline)
-        if iscomplex_X:
+
+        # Write data rows
+        if fstring_fmt:
+            for row in X:
+                if iscomplex_X:
+                    row = [(item.real, item.imag) for item in row]
+                    row = [item for sublist in row for item in sublist]
+                row_formatted = format_row.format(*row)
+                fh.write(row_formatted + newline)
+        elif iscomplex_X:
             for row in X:
-                row2 = []
-                for number in row:
-                    row2.append(number.real)
-                    row2.append(number.imag)
+                row2 = [number.real if i % 2 == 0 else number.imag for i, number in
+                        enumerate(row)]
                 s = format % tuple(row2) + newline
-                fh.write(s.replace('+-', '-'))
+                fh.write(s.replace("+-", "-"))
         else:
             for row in X:
                 try:
                     v = format % tuple(row) + newline
                 except TypeError as e:
-                    raise TypeError("Mismatch between array dtype ('%s') and "
-                                    "format specifier ('%s')"
-                                    % (str(X.dtype), format)) from e
+                    raise TypeError(
+                        f"Mismatch between array dtype ('{X.dtype}') and format specifier ('{format}')"
```

Comment:
This code change can introduce different output. f-string formatting uses the [\_\_format\_\_](https://docs.python.org/3/reference/datamodel.html#object.__format__) protocol. In the original code the `X.dtype` was converted using `str(X.dtype)`. Please revert the changes, they are not related to the fstring_fmt.